lib/screenplay_web/controllers/config_controller.ex

```diff
@@ -94,20 +94,115 @@ defmodule ScreenplayWeb.ConfigController do
+            }
+          },
+          version_id: String.t(),
+          last_modified_ms: integer
+        }
+  defp get_existing_screens_at_places_with_pending_screens do
+    {pending_screens_config, version_id, last_modified} =
+      case PendingScreensConfig.fetch_config() do
+        {:ok, config, version_id, last_modified} ->
+          %PendingConfig{screens: pending_screens} =
+            config
+            |> Jason.decode!()
+            |> PendingConfig.from_json()
+
+          {pending_screens, version_id, last_modified}
+
+        _ ->
+          raise(
+            "Could not fetch pending screens config in existing_screens_at_places_with_pending_screens/2"
+          )
+      end
+
+    existing =
+      pending_screens_config
+      |> Enum.group_by(fn {_, screen} -> {screen_to_place_id(screen), screen.app_id} end)
+      |> Map.new(fn {{place_id, app_id}, pending_screens_at_place} ->
+        filter_fn = fn {_, screen} ->
+          screen.app_id == app_id and screen_to_place_id(screen) == place_id
+        end
+
+        live_screens_of_same_type_at_place =
+          for {id, screen} <- ScreensConfigCache.screens(filter_fn),
+              into: %{},
+              do: {id, Screen.to_json(screen)}
+
+        pending_screens_at_place =
+          for {id, screen} <- pending_screens_at_place,
+              into: %{},
+              do: {id, Screen.to_json(screen)}
+
+        live_and_pending = %{
+          live_screens: live_screens_of_same_type_at_place,
+          pending_screens: pending_screens_at_place,
+          place_id: place_id,
+          app_id: app_id
+        }
+
+        json_key = "#{place_id}/#{app_id}"
+
+        {json_key, live_and_pending}
+      end)
+
+    %{
+      places_and_screens: existing,
+      version_id: version_id,
+      last_modified_ms: DateTime.to_unix(last_modified, :millisecond)
+    }
+  end
+
   def publish(conn, %{
         "place_id" => place_id,
         "app_id" => app_id,
-        "hidden_from_screenplay_ids" => hidden_from_screenplay_ids
+        "hidden_from_screenplay_ids" => hidden_from_screenplay_ids,
+        "version_id" => version_id
       }) do
     app_id_atom = String.to_existing_atom(app_id)
 
-    case PermanentConfig.publish_pending_screens(
-           place_id,
-           app_id_atom,
-           hidden_from_screenplay_ids
-         ) do
-      :ok -> send_resp(conn, 200, "OK")
-      _ -> send_resp(conn, 500, "Could not publish screens")
+    case PendingScreensConfig.fetch_config() do
+      {:ok, _, ^version_id, _} ->
+        case PermanentConfig.publish_pending_screens(
+               place_id,
+               app_id_atom,
+               hidden_from_screenplay_ids
+             ) do
+          :ok -> send_resp(conn, 200, "OK")
+          _ -> send_resp(conn, 500, "Could not publish screens. Please contact an engineer.")
+        end
+
+      {:ok, _, _different_version_id, _} ->
+        send_resp(
```

Comment:
These two seem like the most applicable response codes for this situation!

[428 - if version ID is not included in the request](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/428)
> The origin server requires the request to be conditional. This response is intended to prevent the 'lost update' problem, where a client GETs a resource's state, modifies it and PUTs it back to the server, when meanwhile a third party has modified the state on the server, leading to a conflict.

[412 - if version ID from request does not match that of the current config file](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/412)
> This happens with conditional requests on methods other than `GET` or `HEAD` when the condition defined by the `If-Unmodified-Since` or `If-None-Match` headers is not fulfilled.

---

I'll see if I can put the version ID into an `If-Match` header, since that seems like the HTTP-standard way to do it. Though, the MDN docs for it are confusing--it's labeled as a forbidden header, but doesn't show up in the [actual list of forbidden headers](https://developer.mozilla.org/en-US/docs/Glossary/Forbidden_header_name)...
> For other methods, and in particular for `PUT`, If-Match can be used to prevent the [lost update problem](https://www.w3.org/1999/04/Editing/#3.1). It can check if the modification of a resource that the user wants to upload will not override another change that has been done since the original resource was fetched.

Reply:
TIL! Those actually sound perfect, but I had no idea they existed! If it ends up being more involved than you expect, it might be a good idea to do it as a follow-up task to this one just to help with scope here.